src/hostname/hostnamed.c

```diff
@@ -283,28 +283,42 @@ static const char* fallback_chassis(void) {
         case 1: /* Desktop */
         case 3: /* Workstation */
         case 6: /* Appliance PC */
-                return "desktop";
+                return strdup("desktop");
 
         case 2: /* Mobile */
-                return "laptop";
+                return strdup("laptop");
 
         case 4: /* Enterprise Server */
         case 5: /* SOHO Server */
         case 7: /* Performance Server */
-                return "server";
+                return strdup("server");
 
         case 8: /* Tablet */
-                return "tablet";
+                return strdup("tablet");
 
         default:
                 log_debug("Unhandled ACPI PM profile 0x%02x, ignoring.", t);
         }
 
+try_devicetree:
+        r = read_one_line_file("/sys/firmware/devicetree/base/chassis-type", &type);
+        if (r < 0) {
+                log_debug_errno(r, "Failed read device-tree chassis type, ignoring: %m");
+                return NULL;
+        }
+
+        if (valid_chassis(type)) {
```

Comment:
please change valid_chassis() to always return the constsant string from our list of valid chassis types. i.e. change nulstr_contains to return `const char*` instead of `bool` and then return NULL where previously false was returned and the pointer to the nulstr entry otherwise.

That way the patch will be much simpler, and no memory allocation will be necessary.